Feature/1 initial repository setup

miroslavpojer opened the pull request:
Create the initial version of the project.
Populate by logic from the origin project and integrate.

Closes #1

Zejnilovic commented on .github/workflows/release_notes_comments_migration.yml:
```diff
@@ -0,0 +1,94 @@
+            const regexPattern = /([Cc]los(e|es|ed)|[Ff]ix(es|ed)?|[Rr]esolv(e|es|ed))\s*#\s*([0-9]+)/g;
+            
+            let match;
+            while ((match = regexPattern.exec(description)) !== null) {
```
I have an issue to wrap my head around this one

miroslavpojer replied:
This is a regex detector of the mentioned Issue in the PR description. It does not looks nice, yes.

Zejnilovic commented on .github/workflows/unit_tests.yml:
```diff
@@ -0,0 +1,20 @@
+          pip install -r requirements.txt
+      - name: Execute unit tests
+        run: |
+          pytest ./tests
```
Steal coverage from here - https://github.com/AbsaOSS/datasets-similarity/blob/master/.github/workflows/py_test.yml

miroslavpojer replied:
I stole it. Unit tests are not part of the current solution. It will be solved in following not init Issues.

Zejnilovic commented on .gitignore:
```diff
@@ -158,3 +158,4 @@ cython_debug/
 #  and can be added to the global gitignore or merged into this file.  For a more nuclear
 #  option (not recommended) you can uncomment the following to ignore the entire idea folder.
 #.idea/
+/.idea/
```
pick one

miroslavpojer replied:
Addressed in commit - bcb84a552552cb710f87486bbee433af62b38f6d.

Zejnilovic commented on README.md:
```diff
@@ -1,2 +1,127 @@
+We welcome contributions to the Living Documentation Generator! Whether you're fixing bugs, improving documentation, or proposing new features, your help is appreciated.
+
+#### How to Contribute
+- **Submit Pull Requests**: Feel free to fork the repository, make changes, and submit a pull request. Please ensure your code adheres to the existing style and all tests pass.
+- **Report Issues**: If you encounter any bugs or issues, please report them via the repository's [Issues page](https://github.com/AbsaOSS/living-doc-generator/issues).
+- **Suggest Enhancements**: Have ideas on how to make this action better? Open an issue to suggest enhancements.
+
+Before contributing, please review our [contribution guidelines](https://github.com/AbsaOSS/living-doc-generator/blob/master/CONTRIBUTING.md) for more detailed information.
```
```suggestion
Before contributing, please review our [contribution guidelines](https://github.com/AbsaOSS/living-doc-generator/blob/master/CONTRIBUTING.md) for more detailed information.
```

miroslavpojer replied:
Addressed in commit - bcb84a552552cb710f87486bbee433af62b38f6d.

miroslavpojer commented on .github/workflows/release_notes_comments_migration.yml:
```diff
@@ -0,0 +1,94 @@
+name: Copy Release Notes to Related Issues
+
+on:
+  pull_request:
+    types: [closed]
+    branches: [ master ]
+
+jobs:
+  copy_release_notes:
+    if: github.event.pull_request.merged == true
```
The logic is to migrate all Release notes comments from PR to related Issues only when PR is closed.
By push, there would be repeated duplication.

miroslavpojer commented on .github/workflows/check_pr_release_note_comment.yml:
```diff
@@ -0,0 +1,43 @@
+              core.setFailed('No "Release notes" found in PR comments')
+            } else {
+              console.log('"Release notes" found in comments');
+            }
```
Fixed in commit - 0c435b7ef30e44f9f87835d376ea122c58939207.

miroslavpojer commented on .github/workflows/release_draft.yml:
```diff
@@ -0,0 +1,115 @@
+        uses: actions/github-script@v7
+        with:
+          script: |
+            const newTag = core.getInput('tag-name');
```
Fixed in commit - 0c435b7ef30e44f9f87835d376ea122c58939207.

miroslavpojer commented on .github/workflows/release_notes_comments_migration.yml:
```diff
@@ -0,0 +1,94 @@
+                  body: commentBody
+                });            
+              }
+            }
```
Fixed in commit - 0c435b7ef30e44f9f87835d376ea122c58939207.

miroslavpojer commented on .github/workflows/release_notes_comments_migration.yml:
```diff
@@ -0,0 +1,94 @@
+        run: |
+          echo "Extracted Release Note Body:"
+          echo "${{ steps.get-comments.outputs.releaseNoteBody }}"
+          echo "RELEASE_NOTE_BODY<<EOF" >> $GITHUB_ENV
```
Where?

miroslavpojer commented:
Release notes:
- Empty project populated by all required project logic stuff.
- Introduced Python code solution for data mining from GH repositories and Projects
- Introduced GH Action logic to provide Python code to be used by GH Action usage.
- Usage of action tested in example project.

MobiTikula commented on README.md:
````diff
@@ -1,2 +1,299 @@
+  - **Description**: This output provides the path to the directory where the generated living documentation files are stored.
+  - **Usage**: 
+   ``` yaml
+    - name: Generate Living Documentation
````
Change to `Living Documentation Generator`.

MobiTikula commented on README.md:
````diff
@@ -1,2 +1,299 @@
+See the default action step definition:
+
+```yaml
+- name: Generate Living Documentation
````
May change name to `Generate Living Documentation` everywhere in README as it is in action.yml file.

miroslavpojer replied:
I have checked the file and fixed capital in 1st line. Other variations in the file still make sense to me.
Please provide reasons for the change in local meanings.

MobiTikula commented on dist/controller.py:
```diff
@@ -0,0 +1,75 @@
+import re
+
+
+def extract_args():
```
Type hints and proper method / script documentation has to be added to the whole `controller.py`.

miroslavpojer replied:
Please address this in the SPIKE issue as one of the goals.

MobiTikula commented on dist/controller.py:
```diff
@@ -0,0 +1,75 @@
+    run_script('convert_features_to_pages.py', env_vars)
+
+
+if __name__ == '__main__':
```
Here is different pattern, than we usually have in our scripts, where we don't initialize main method. I would refactor the code so the scripts are consistent.

miroslavpojer replied:
Please address this in the SPIKE issue as one of the goals.
This version aimed to provide a migrated solution that worked well enough to close previous projects.
No deep analysis was included, as this step prevented us from developing the final project.

lsulak commented on src/controller.py:
```diff
@@ -0,0 +1,75 @@
+
+if __name__ == '__main__':
+    print("Starting Living Documentation generation.")
+    main()
```
This PR is seriously very big, almost no one will be able to review it fully in detail within reasonable amount of time. Better would be to create a few smaller ones. In your case, in particular, you could add one script at a time. One PR for `clean_env_before_mining.py` and related functionality, another for `github_query_issues.py` etc.

miroslavpojer replied:
This version aimed to provide a migrated solution that worked well enough to close previous projects.
The Python scripts are the goal of that migration. We are working on collecting refactoring notes in parallel.
- in a previous project
- in a local project in SPIKE issue

Yes, it is big, but the script is the same. Only the main method and env variables loading differ.
All scripts need to be refactored. Adding a format check and using OOP will be the first steps; then, more reviews should follow.

lsulak commented on src/controller.py:
```diff
@@ -0,0 +1,75 @@
+        env.update(env_vars)
+
+        # Running the script with updated environment
+        result = subprocess.run(['python3', script_name], env=env, text=True, capture_output=True, check=True)
```
calling Python from Python using subprocess is definitely not that common. Why not to call it directly? If env vars are the only reason, you can encapsulate them into a data class (I proposed something like that already in: https://github.com/AbsaOSS/living-doc-generator/issues/4#issuecomment-2115300899) and pass to each main function of each module, just an example.

miroslavpojer replied:
@MobiTikula Please address it in SPIKE for refactoring.

miroslavpojer replied:
See the solution proposal linked on following review comment.

lsulak commented on src/controller.py:
```diff
@@ -0,0 +1,75 @@
+
+    # Clean environment before mining
+    print("Data mining for Living Documentation")
+    run_script('clean_env_before_mining.py', env_vars)
```
hardcoding python script name like this is definitely weird. I'd just import and call it directly as a standard Python call

miroslavpojer replied:
I have prepared a solution proposal for the clean script - see commit 56c6fc915d1a0f570f50306754a81381df6db610.
@MobiTikula, Can you address the refactoring of the rest of the scripts in SPIKE?

lsulak commented on src/consolidate_feature_data.py:
```diff
@@ -0,0 +1,239 @@
+    # Set to store the names of repositories that have been used
+    set_of_used_repos = set()
+
+    if os.path.isdir(PROJECT_DIRECTORY):
```
this function is quite nested and I feel like almost a bit too long. Consider simplifying it and perhaps even break it down into multiple functions.

Btw, I like this approach to 'flatten' the code a bit - instead of:

```python
if something:
   do_something_complex_and_nested
```

to:

```python
if not something:
   return

do_something_complex_and_nested
```

by following this second approach, your `do_something_complex_and_nested` code block is not nested any more :)

lsulak replied:
This is cool: https://peps.python.org/pep-0020/

`Flat is better than nested.`

miroslavpojer replied:
As this was overtaken by migration, let's make this a note to refactor SPIKE @MobiTikula .
